058-zk-kraft-migration.md

```diff
@@ -0,0 +1,433 @@
+
+The ZooKeeper to KRaft migration implementation is defined by [KIP-866](https://cwiki.apache.org/confluence/display/KAFKA/KIP-866+ZooKeeper+to+KRaft+Migration).
+The upstream Apache Kafka community documentation also provides a [ZooKeeper to KRaft Migration](https://kafka.apache.org/documentation/#kraft_zk_migration) guide showing how to run the procedure manually on your cluster not deployed via the Strimzi Cluster Operator.
+
+One of the main points to highlight is that such a process doesn't support migration to a KRaft-based cluster running in "combined" mode:
+
+* Cannot migrate from a ZooKeeper based configuration to a cluster with nodes hosting both controller and broker (so actually reducing the number of nodes by removing the ZooKeeper ones).
+* The migration works by having ZooKeeper nodes to be replaced by new KRaft controller nodes.
+* Supporting "combined" mode is actually one of the rejected [alternatives](https://cwiki.apache.org/confluence/display/KAFKA/KIP-866+ZooKeeper+to+KRaft+Migration#KIP866ZooKeepertoKRaftMigration-CombinedModeMigrationSupport) in KIP-866.
+
+It means that the user environment should be sized to host more nodes running in parallel, because during the migration process both ZooKeeper nodes and KRaft controllers will be running.
+The end of the migration will involve shutting down the ZooKeeper nodes reducing the resources consumption.
+
+After the migration, the users should be also able to migrate to use combined nodes for an additional reduction of resources, but that is out of the scope for this proposal.
+
+## Motivation
+
+The KRaft mode has been declared to be production-ready since Apache Kafka 3.3 release by the Apache Kafka project. 
+Anyway, there are still missing features as reported in the official [documentation](https://kafka.apache.org/documentation/#kraft_missing).
+The ZooKeeper mode has been deprecated after the Apache Kafka 3.5 version was released while providing the migration process to KRaft mode as preview.
+The coming Apache Kafka 3.6 release will have such a migration process supported as GA and the ZooKeeper mode will be supported until Apache Kafka 3.7 but removed after that.
+Starting from Apache Kafka 4.0 only KRaft mode will be supported.
+
+Based on the above Apache Kafka community plan, more details in [KIP-833](https://cwiki.apache.org/confluence/display/KAFKA/KIP-833%3A+Mark+KRaft+as+Production+Ready), the Strimzi project needs to be prepared to allow users to migrate their current clusters from running in ZooKeeper mode to KRaft mode.
+
+## Proposal
+
+### Prerequisites
+
+In order to start the ZooKeeper to KRaft migration, the following prerequisites should be met:
+
+* A Strimzi-managed Apache Kafka cluster running in ZooKeeper mode.
```

Comment:
I am not sure about this. The Kafka upstream community should aim to have a KRaft support with no missing features, or?

Reply:
Yes but ZK is already deprecated, so actually no one knows for sure what will happen. Even if this in unlikely, there is no written agreement on that. Anyway, we can leave that out and use documentation in case of any missing feature.